semantic-analysis/src/Analysis/Syntax/Python.hs

```diff
@@ -0,0 +1,198 @@
+  -> [T.Term Python v]
+  -> T.Term Python v
+pattern Call f as <- Call' T.:$: T.Cons f (T.Cons (fromArgs -> as) T.Nil)
+  where Call f as = Call' T.:$: T.Cons f (T.Cons (foldr ACons ANil as) T.Nil)
+
+fromArgs :: T.Term Python v -> [T.Term Python v]
+fromArgs = \case
+  ANil -> []
+  ACons a as -> a:fromArgs as
+  _ -> fail "unexpected constructor in spine of argument list"
+
+pattern ANil :: T.Term Python v
+pattern ANil = ANil' T.:$: T.Nil
+
+pattern ACons :: T.Term Python v -> T.Term Python v -> T.Term Python v
+pattern ACons a as = ACons' T.:$: T.Cons a (T.Cons as T.Nil)
+
+pattern Locate :: Span -> T.Term Python v -> T.Term Python v
+pattern Locate s t = Locate' s T.:$: T.Cons t T.Nil
+
+{-# COMPLETE Noop, Iff, Bool, String, Throw, Let, (:>>), Import, Function, Call, Locate #-}
+
+
+instance Eq1 Python where
+  liftEq _ a b = case (a, b) of
+    (Noop', Noop')                       -> True
+    (Iff', Iff')                         -> True
+    (Bool' b1, Bool' b2)                 -> b1 == b2
+    (String' s1, String' s2)             -> s1 == s2
+    (Throw', Throw')                     -> True
+    (Let' n1, Let' n2)                   -> n1 == n2
+    ((:>>>), (:>>>))                     -> True
+    (Import' i1, Import' i2)             -> i1 == i2
+    (Function' n1 as1, Function' n2 as2) -> n1 == n2 && as1 == as2
+    (Call', Call')                       -> True
+    (ANil', ANil')                       -> True
+    (ACons', ACons')                     -> True
+    (Locate' s1, Locate' s2)             -> s1 == s2
+    _                                    -> False
+
+instance Ord1 Python where
+  liftCompare _ a b = case (a, b) of
+    (Noop', Noop')                       -> EQ
+    (Noop', _)                           -> LT
+    (Iff', Iff')                         -> EQ
+    (Iff', _)                            -> LT
+    (Bool' b1,  Bool' b2)                -> compare b1 b2
+    (Bool' _,  _)                        -> LT
+    (String' s1, String' s2)             -> compare s1 s2
+    (String' _, _)                       -> LT
+    (Throw', Throw')                     -> EQ
+    (Throw', _)                          -> LT
+    (Let' n1, Let' n2)                   -> compare n1 n2
+    (Let' _, _)                          -> LT
+    ((:>>>), (:>>>))                     -> EQ
+    ((:>>>), _)                          -> LT
+    (Import' i1, Import' i2)             -> compare i1 i2
+    (Import' _, _)                       -> LT
+    (Function' n1 as1, Function' n2 as2) -> compare n1 n2 <> compare as1 as2
+    (Function' _ _, _)                   -> LT
+    (Call', Call')                       -> EQ
+    (Call', _)                           -> LT
+    (ANil', ANil')                       -> EQ
+    (ANil', _)                           -> LT
+    (ACons', ACons')                     -> EQ
+    (ACons', _)                          -> LT
+    (Locate' s1, Locate' s2)             -> compare s1 s2
+    (Locate' _, _)                       -> LT
```

Comment:
We can't derive these because `GHC.Generics` doesn't play well with `GADTs` (or the other way around, I suppose). This is substantially worse than what we'd get using standard `Fix`-based terms or even plain old data.